src/FxKit.CompilerServices/CodeGenerators/Transformers/Functor.cs

```diff
@@ -1,175 +1,238 @@
-﻿using Microsoft.CodeAnalysis;
+﻿using System.Diagnostics;
+using FxKit.CompilerServices.Utilities;
+using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
 using static FxKit.CompilerServices.CodeGenerators.Transformers.Helpers;
 
 namespace FxKit.CompilerServices.CodeGenerators.Transformers;
 
 /// <summary>
-///     Represents an outer container functor type.
+///     An extraction from a referenced assembly.
 /// </summary>
-public readonly struct Functor
+/// <param name="Candidates">The found functors.</param>
+/// <param name="Locators">
+///     Potential locators; they have not been verified yet as we need to know all functors first.
+/// </param>
+internal sealed record ReferencedFunctors(
+    EquatableArray<FunctorCandidate> Candidates,
+    EquatableArray<FunctorImplementationLocator> Locators)
 {
     /// <summary>
-    ///     The name of the functor (its identifier).
+    ///     Finds functors and functor implementations in the given assembly reference.
     /// </summary>
-    public string Name { get; }
+    /// <remarks>
+    ///     We're extracting both functors (typed with a `[Functor]` attribute), as well
+    ///     as functor implementations (`Map` methods) in a single pass to avoid having
+    ///     to iterate on references multiple times.
+    /// </remarks>
+    /// <param name="reference"></param>
+    /// <param name="compilation"></param>
+    /// <param name="cancellationToken"></param>
+    /// <returns></returns>
+    public static ReferencedFunctors? GetFunctorsAndBehaviorFromReference(
+        MetadataReference reference,
+        Compilation compilation,
+        CancellationToken cancellationToken)
+    {
+        if (compilation.GetAssemblyOrModuleSymbol(reference) is not IAssemblySymbol asm)
+        {
+            return null;
+        }
+
+        List<FunctorCandidate>? candidates = null;
+        List<FunctorImplementationLocator>? locators = null;
+        var typeSymbols = TypeSymbolHelper.GetTypesRecursive(asm.GlobalNamespace, cancellationToken);
+        foreach (var typeSymbol in typeSymbols)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            foreach (var attributeData in typeSymbol.GetAttributes())
+            {
+                if (attributeData.AttributeClass?.GetFullyQualifiedMetadataName() !=
+                    TransformerGenerator.FunctorAttribute)
+                {
+                    continue;
+                }
+
+                candidates ??= [];
+                candidates.Add(FunctorCandidate.From(typeSymbol));
+            }
+
+            var methodSymbols = typeSymbol.GetMembers("Map")
```

Comment:
It's not important, but we could raise `"Map"` to a static `FunctorMethodName` field or equivalent.

Reply:
Good idea.